### PR TITLE
Implement SynProxy

### LINKED ISSUE #15: json schema is missing synproxy support
some info about synproxy can be found here https://wiki.nftables.org/wiki-nftables/index.php/Synproxy

The field is not documented in the libnftables-json(5) page

some example output for named synproxy
```
# nft list ruleset
table ip foo {
	synproxy https-synproxy {
		mss 1460
		wscale 7
		timestamp sack-perm
	}

	synproxy other-synproxy {
		mss 1460
		wscale 5
	}

	chain pre {
		type filter hook prerouting priority raw; policy accept;
		tcp dport 8888 tcp flags syn notrack
	}

	chain bar {
		type filter hook forward priority filter; policy accept;
		ct state invalid,untracked synproxy name ip saddr map { 192.168.1.0/24 : "https-synproxy", 192.168.2.0/24 : "other-synproxy" }
	}
}

# nft -j list ruleset | jq
{
  "nftables": [
    {
      "metainfo": {
        "version": "1.0.7",
        "release_name": "Old Doc Yak",
        "json_schema_version": 1
      }
    },
    {
      "table": {
        "family": "ip",
        "name": "foo",
        "handle": 2
      }
    },
    {
      "synproxy": {
        "family": "ip",
        "name": "https-synproxy",
        "table": "foo",
        "handle": 3,
        "mss": 1460,
        "wscale": 7,
        "flags": [
          "timestamp",
          "sack-perm"
        ]
      }
    },
    {
      "synproxy": {
        "family": "ip",
        "name": "other-synproxy",
        "table": "foo",
        "handle": 4,
        "mss": 1460,
        "wscale": 5
      }
    },
    {
      "chain": {
        "family": "ip",
        "table": "foo",
        "name": "pre",
        "handle": 1,
        "type": "filter",
        "hook": "prerouting",
        "prio": -300,
        "policy": "accept"
      }
    },
    {
      "chain": {
        "family": "ip",
        "table": "foo",
        "name": "bar",
        "handle": 2,
        "type": "filter",
        "hook": "forward",
        "prio": 0,
        "policy": "accept"
      }
    },
    {
      "rule": {
        "family": "ip",
        "table": "foo",
        "chain": "pre",
        "handle": 5,
        "expr": [
          {
            "match": {
              "op": "==",
              "left": {
                "payload": {
                  "protocol": "tcp",
                  "field": "dport"
                }
              },
              "right": 8888
            }
          },
          {
            "match": {
              "op": "in",
              "left": {
                "payload": {
                  "protocol": "tcp",
                  "field": "flags"
                }
              },
              "right": "syn"
            }
          },
          {
            "notrack": null
          }
        ]
      }
    },
    {
      "rule": {
        "family": "ip",
        "table": "foo",
        "chain": "bar",
        "handle": 7,
        "expr": [
          {
            "match": {
              "op": "in",
              "left": {
                "ct": {
                  "key": "state"
                }
              },
              "right": [
                "invalid",
                "untracked"
              ]
            }
          },
          {
            "synproxy": {
              "map": {
                "key": {
                  "payload": {
                    "protocol": "ip",
                    "field": "saddr"
                  }
                },
                "data": {
                  "set": [
                    [
                      {
                        "prefix": {
                          "addr": "192.168.1.0",
                          "len": 24
                        }
                      },
                      "https-synproxy"
                    ],
                    [
                      {
                        "prefix": {
                          "addr": "192.168.2.0",
                          "len": 24
                        }
                      },
                      "other-synproxy"
                    ]
                  ]
                }
              }
            }
          }
        ]
      }
    }
  ]
}

```

and for anonymous synproxy

```
# nft list ruleset
table ip anon_synproxy_demo {
	chain PRE {
		type filter hook prerouting priority raw; policy accept;
		tcp dport 8888 tcp flags syn notrack
	}

	chain IN {
		type filter hook input priority filter; policy accept;
		tcp dport 8888 ct state invalid,untracked synproxy mss 1460 wscale 7 timestamp sack-perm
		ct state invalid drop
	}
}
root@pholzing-fedora:/home/pholzing/CODE/nftables-rs# nft -j list ruleset | jq
{
  "nftables": [
    {
      "metainfo": {
        "version": "1.0.7",
        "release_name": "Old Doc Yak",
        "json_schema_version": 1
      }
    },
    {
      "table": {
        "family": "ip",
        "name": "anon_synproxy_demo",
        "handle": 3
      }
    },
    {
      "chain": {
        "family": "ip",
        "table": "anon_synproxy_demo",
        "name": "PRE",
        "handle": 1,
        "type": "filter",
        "hook": "prerouting",
        "prio": -300,
        "policy": "accept"
      }
    },
    {
      "chain": {
        "family": "ip",
        "table": "anon_synproxy_demo",
        "name": "IN",
        "handle": 2,
        "type": "filter",
        "hook": "input",
        "prio": 0,
        "policy": "accept"
      }
    },
    {
      "rule": {
        "family": "ip",
        "table": "anon_synproxy_demo",
        "chain": "PRE",
        "handle": 3,
        "expr": [
          {
            "match": {
              "op": "==",
              "left": {
                "payload": {
                  "protocol": "tcp",
                  "field": "dport"
                }
              },
              "right": 8888
            }
          },
          {
            "match": {
              "op": "in",
              "left": {
                "payload": {
                  "protocol": "tcp",
                  "field": "flags"
                }
              },
              "right": "syn"
            }
          },
          {
            "notrack": null
          }
        ]
      }
    },
    {
      "rule": {
        "family": "ip",
        "table": "anon_synproxy_demo",
        "chain": "IN",
        "handle": 4,
        "expr": [
          {
            "match": {
              "op": "==",
              "left": {
                "payload": {
                  "protocol": "tcp",
                  "field": "dport"
                }
              },
              "right": 8888
            }
          },
          {
            "match": {
              "op": "in",
              "left": {
                "ct": {
                  "key": "state"
                }
              },
              "right": [
                "invalid",
                "untracked"
              ]
            }
          },
          {
            "synproxy": {
              "mss": 1460,
              "wscale": 7,
              "flags": [
                "timestamp",
                "sack-perm"
              ]
            }
          }
        ]
      }
    },
    {
      "rule": {
        "family": "ip",
        "table": "anon_synproxy_demo",
        "chain": "IN",
        "handle": 5,
        "expr": [
          {
            "match": {
              "op": "in",
              "left": {
                "ct": {
                  "key": "state"
                }
              },
              "right": "invalid"
            }
          },
          {
            "drop": null
          }
        ]
      }
    }
  ]
}
```

found originally in https://github.com/containers/netavark/issues/942

### PR DESCRIPTION
**Pull request by jwhb**
SynProxy was missing, as it is not mentioned in the libnftables-json man page.

A few other statements were found to be implement in libnftables-json while not being documented by libnftables. TODO marks have been added for these statements.

Fixes #15.

### CODE REVIEW

File: src/stmt.rs [hunk 1]
```diff
@@ -48,6 +50,7 @@ pub enum Statement {
     Redirect(Option<NAT>),   // redirect is subset of NAT options
     Reject(Option<Reject>),
     Set(Set),
+    // TODO: map
```
Comment (JKRhb) [inline src/stmt.rs:53]: Hmm, should these TODOs stay in the codebase? Or should we maybe rather create an issue with a task list for these? (Or both?)

File: src/stmt.rs [hunk 2]
```diff
@@ -36,8 +36,10 @@ pub enum Statement {
     #[serde(rename = "quota")]
```
Comment (JKRhb) [inline src/stmt.rs:36]: If there are more entries that are going to be added in the future, should this enum maybe be marked as `non_exhaustive`?

File: src/stmt.rs [hunk 3]
```diff
@@ -432,6 +439,19 @@ pub struct CTCount {
...
+/// Limit the number of connections using conntrack.
+pub struct SynProxy {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// maximum segment size (must match your backend server)
+    pub mss: Option<u32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// window scale (must match your backend server)
+    pub wscale: Option<u32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub flags: Option<HashSet<SynProxyFlag>>,
```
Comment (JKRhb) [inline src/stmt.rs:452]: There could be another doc comment added here.

File: src/schema.rs [hunk 4]
```diff
@@ -435,3 +436,20 @@ pub struct CTExpectation {
     #[serde(skip_serializing_if = "Option::is_none")]
     pub size: Option<u32>,
 }
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct SynProxy {
+    pub family: NfFamily,
+    pub table: String,
+    pub name: String,
```
Comment (JKRhb) [inline src/schema.rs:444]: Should these entries also be documented?

File: src/schema.rs [hunk 5]
```diff
@@ -435,3 +436,20 @@ pub struct CTExpectation {
...
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub handle: Option<u32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// maximum segment size (must match your backend server)
+    pub mss: Option<u16>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// window scale (must match your backend server)
+    pub wscale: Option<u8>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub flags: Option<HashSet<SynProxyFlag>>,
```
Comment (JKRhb) [inline src/schema.rs:454]: Could also be documented.

**Review by JKRhb — COMMENTED** [review]
See a couple of comments below :)

File: src/schema.rs (hunk 4 above)
Comment (jwhb) [reply src/schema.rs]: Thanks, fixed in 855b72b8dfb9a3010a84c71c290a724e9a75743c

File: src/schema.rs (hunk 5 above)
Comment (jwhb) [reply src/schema.rs:465]: Thanks, fixed in 855b72b8dfb9a3010a84c71c290a724e9a75743c

File: src/stmt.rs (hunk 3 above)
Comment (jwhb) [reply src/stmt.rs:455]: Thanks, fixed in 855b72b8dfb9a3010a84c71c290a724e9a75743c

File: src/schema.rs (hunk 4 above)
Comment (jwhb) [reply src/schema.rs]: I'd prefer to keep the TODO marks in the code base for future code readers to be aware of this.

Created an issue: #70

File: src/stmt.rs (hunk 1 above)
Comment (jwhb) [reply src/stmt.rs:53]: I prefer to keep the marks in the code base for future code readers to be aware of this.

Created an issue: #70.

File: src/stmt.rs (hunk 2 above)
Comment (jwhb) [reply src/stmt.rs:36]: Good point. Addressed in #70 and I think that `non-exhaustive` should be added in v0.5.0.

**Review by JKRhb — APPROVED** [review]
LGTM! :)

**Pull request merged**